File: 1D DP/000 Pracitce/Subset Sum/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
    public:
    bool subsetsum(const vector<int>& arr, int target){
        int n = arr.size();
        vector<int> dp(target+1);
        dp[0] = 1;

        for (int i=0; i<n; i++){
            for (int j=target; j>=arr[i]; j--){
                int num = arr[i];
                dp[j] = dp[j-num] || dp[j];
            }

            if (dp[target]) return true;
        }
        return dp[target]; 
    }
};

int main(){
    
    Solution solution;

    vector<int> arr = {3, 34, 4, 12, 5, 2};
    int sum = 9;

    cout << solution.subsetsum(arr, sum);
    return 0;
}

File: 1D DP/000 Pracitce/Subset Sum/main.py
class Solution:
    @staticmethod
    def subsetsum(arr:list[int], target:int) -> bool:
        n = len(arr)
        dp = [False] * (target + 1)
        dp[0] = True

        for i in range(n):
            for j in range(target, arr[i] - 1, -1):
                dp[j] = dp[j - arr[i]] or dp[j]

            if dp[target]:
                return True

        return dp[target]
    
print(Solution.subsetsum(
    arr=[1,2,5],
    target = 4
))